### PR TITLE
Added string operations support

### PR DESCRIPTION
**Pull request by ymoran00**
Added support for aggregation operators - all string operations that weren't supported yet, and a new $toString operation that was added in mongodb 4.0

### CODE REVIEW

File: mongomock/aggregate.py [hunk 1]
```diff
@@ -87,6 +87,9 @@
     '$toLower',
     '$toUpper',
 ]
+misc_operators = [
+    '$toString',
```
Comment (pcorpet) [inline mongomock/aggregate.py:90]: Let's add it to the string operators (the string operations convert non-strings to string already).

File: mongomock/aggregate.py [hunk 2]
```diff
@@ -157,7 +160,11 @@ def parse(self, expression):
                 return self._handle_conditional_operator(k, v)
             if k in set_operators:
                 return self._handle_set_operator(k, v)
-            if k in boolean_operators + string_operators + \
+            if k in string_operators:
+                return self._handle_string_operator(k, v)
+            if k in misc_operators:
+                return self._handle_misc_operator(k, v)
+            if k in boolean_operators + \
```
Comment (pcorpet) [inline mongomock/aggregate.py:164]: Please update the Missing_features.rst file to mention that those features are not missing anymore.

File: tests/test__collection_api.py [hunk 3]
```diff
@@ -2597,6 +2597,36 @@ def test__aggregate_arithmetic(self):
             [{'sum': 20.5, 'prod': 30, 'trunc': 1}],
             [{k: v for k, v in doc.items() if k != '_id'} for doc in actual])
 
+    def test__aggregate_string_operations(self):
+        self.db.collection.insert_one({
+            'a': 'Hello',
+            'b': 'World',
+        })
+        actual = self.db.collection.aggregate([{'$project': {
```
Comment (pcorpet) [inline tests/test__collection_api.py:2616]: Please also add that test to test__mongomock.py to make sure the behavior is consistant with Pymongo.

File: tests/test__collection_api.py [hunk 4]
```diff
@@ -2597,6 +2597,36 @@ def test__aggregate_arithmetic(self):
...
+            'sub2': {'$substr': ['$a', -1, 3]},
+            'lower': {'$toLower': '$a'},
+            'upper': {'$toUpper': '$a'},
+        }}])
+        self.assertEqual(
+            [{'concat': 'Hello Dear World', 'sub1': 'Hell', 'sub2': '', 'lower': 'hello', 'upper': 'HELLO'}],
+            [{k: v for k, v in doc.items() if k != '_id'} for doc in actual])
+
+    def test__aggregate_misc_operations(self):
+        _id = ObjectId()
```
Comment (pcorpet) [inline tests/test__collection_api.py]: Please use an ObjectId for which you choose the hash:
```py
_id = ObjectId('5abcfad1fbc93d00080cfe66')

...

assertEqual([{'toString1: '5abcfad1fbc93d00080cfe66', 'toString2': '3'}], ...)
```

Comment (pcorpet) [reply tests/test__collection_api.py]: The rationale is that I prefer to keep the logic in tests to their minimum.

File: tests/test__collection_api.py (hunk 3 above)
Comment (pcorpet) [reply tests/test__collection_api.py:2616]: Same for the test below.

File: mongomock/aggregate.py [hunk 5]
```diff
@@ -256,6 +263,40 @@ def _handle_comparison_operator(self, operator, values):
...
+
+    def _handle_string_operator(self, operator, values):
+        if operator == '$toLower':
+            return str(self.parse(values)).lower()
+        if operator == '$toUpper':
+            return str(self.parse(values)).upper()
+        if operator == '$concat':
+            return ''.join([str(self.parse(value)) for value in values])
+        if operator == '$substr':
+            assert len(values) == 3, 'substr must have 3 items'
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please be consistant with the error that is raised by pymongo. It usually raises OperationFailure instead of assertion errors.

Same below.

File: mongomock/aggregate.py [hunk 6]
```diff
@@ -256,6 +263,40 @@ def _handle_comparison_operator(self, operator, values):
...
+            string = str(self.parse(values[0]))
+            first = self.parse(values[1])
+            length = self.parse(values[2])
+            if string is None or first < 0:
+                return ''
+            second = len(string) if length < 0 else first + length
```
Comment (pcorpet) [inline mongomock/aggregate.py:288]: Please add a test case for negative length.

File: mongomock/aggregate.py [hunk 7]
```diff
@@ -256,6 +263,40 @@ def _handle_comparison_operator(self, operator, values):
...
+            parsed = self.parse(values)
+            return str(parsed) if parsed is not None else None
+            # This should never happen: it is only a safe fallback if something went wrong.
+        raise NotImplementedError(  # pragma: no cover
+            "Although '%s' is a valid operator for the aggregation "
+            'pipeline, it is currently not implemented  in Mongomock.' % operator)
+
+    def _handle_string_operator(self, operator, values):
+        if operator == '$toLower':
+            return str(self.parse(values)).lower()
```
Comment (pcorpet) [inline mongomock/aggregate.py]: This would fail if values returns None, please add a test for that.

**Review by pcorpet — CHANGES_REQUESTED** [review]
Thanks Yoav, very much appreciated. I have a few comments to keep your code to the existing standards.

File: tests/test__collection_api.py (hunk 3 above)
Comment (ymoran00) [reply tests/test__collection_api.py:2616]: the $toString is only supported since mongodb 4.0. Should it be there as well anyway?

File: mongomock/aggregate.py (hunk 1 above)
Comment (ymoran00) [reply mongomock/aggregate.py:90]: np

File: mongomock/aggregate.py (hunk 2 above)
Comment (ymoran00) [reply mongomock/aggregate.py:164]: np

File: tests/test__collection_api.py (hunk 3 above)
Comment (ymoran00) [reply tests/test__collection_api.py:2616]: Added for now - unless you'll say otherwise.

File: tests/test__collection_api.py (hunk 4 above)
Comment (ymoran00) [reply tests/test__collection_api.py]: good idea. np

File: mongomock/aggregate.py (hunk 5 above)
Comment (ymoran00) [reply mongomock/aggregate.py]: np

File: mongomock/aggregate.py (hunk 7 above)
Comment (ymoran00) [reply mongomock/aggregate.py]: Apparently according to MongoDB doc - it should return '' for None. So I updated the code and added a test accordingly.

File: mongomock/aggregate.py (hunk 6 above)
Comment (ymoran00) [reply mongomock/aggregate.py:288]: forgot it on push - so added the test and pushed again.

File: mongomock/aggregate.py [hunk 8]
```diff
@@ -256,6 +259,38 @@ def _handle_comparison_operator(self, operator, values):
...
 
+    def _handle_string_operator(self, operator, values):
+        if operator == '$toLower':
+            parsed = self.parse(values)
+            return str(parsed).lower() if parsed is not None else ''
+        if operator == '$toUpper':
+            parsed = self.parse(values)
+            return str(parsed).upper() if parsed is not None else ''
+        if operator == '$concat':
+            return ''.join([str(self.parse(value)) for value in values])
```
Comment (pcorpet) [inline mongomock/aggregate.py]: What if one of those subvalue is None? would it crash? Please add a test for this.

File: tests/test__collection_api.py (hunk 3 above)
Comment (pcorpet) [reply tests/test__collection_api.py:2616]: Oh, you're right, Travis does not have MongoDB 4 yet. Please add a TODO there to reenabe once Travis runs Mongo v4+.

Still add the test for the string operators.

File: mongomock/aggregate.py (hunk 8 above)
Comment (ymoran00) [reply mongomock/aggregate.py]: You're right - that's another case. According to docs this will return a concat result of None. Fixed and added test.

File: tests/test__collection_api.py (hunk 3 above)
Comment (ymoran00) [reply tests/test__collection_api.py:2616]: np

File: mongomock/aggregate.py [hunk 9]
```diff
@@ -256,6 +259,40 @@ def _handle_comparison_operator(self, operator, values):
...
+            return str(parsed).lower() if parsed is not None else ''
+        if operator == '$toUpper':
+            parsed = self.parse(values)
+            return str(parsed).upper() if parsed is not None else ''
+        if operator == '$concat':
+            parsed_list = [self.parse(value) for value in values]
+            return None if None in parsed_list else ''.join([str(x) for x in parsed_list])
+        if operator == '$substr':
+            if len(values) != 3:
+                raise OperationFailure('substr must have 3 items')
```
Comment (pcorpet) [inline mongomock/aggregate.py:275]: Please add a test for this.

File: mongomock/aggregate.py [hunk 10]
```diff
@@ -256,6 +259,40 @@ def _handle_comparison_operator(self, operator, values):
...
+            first = self.parse(values[1])
+            length = self.parse(values[2])
+            if string is None or first < 0:
+                return ''
+            second = len(string) if length < 0 else first + length
+            return string[first:second]
+        if operator == '$strcasecmp':
+            if len(values) != 2:
+                raise OperationFailure('strcasecmp must have 2 items')
+            a, b = str(self.parse(values[0])), str(self.parse(values[1]))
```
Comment (pcorpet) [inline mongomock/aggregate.py:293]: Add two test for stcasecmp, one for this, and one for when it succeeds and one for when it fails.

File: mongomock/aggregate.py [hunk 11]
```diff
@@ -256,6 +259,40 @@ def _handle_comparison_operator(self, operator, values):
...
+            return 0 if a == b else -1 if a < b else 1
+        # TODO(ymoran00): Re-enable once TRAVIS supports MongoDB 4
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Sorry, I meant that you should add this TODO on the test in test__mongomock: the feature is welcome and can be tested in test__collection_api

File: tests/test__collection_api.py [hunk 12]
```diff
@@ -2597,6 +2597,41 @@ def test__aggregate_arithmetic(self):
...
+            'lower_err': {'$toLower': None},
+            'upper': {'$toUpper': '$a'},
+            'upper_err': {'$toUpper': None},
+        }}])
+        self.assertEqual(
+            [{'concat': 'Hello Dear World', 'concat_none': None, 'sub1': 'Hell', 'sub2': '',
+              'sub3': 'llo', 'lower': 'hello', 'lower_err': '', 'upper': 'HELLO', 'upper_err': ''}],
+            [{k: v for k, v in doc.items() if k != '_id'} for doc in actual])
+
+    # TODO(ymoran00): Re-enable once TRAVIS supports MongoDB 4
```
Comment (pcorpet) [inline tests/test__collection_api.py]: Keep this for now.

Comment (ymoran00) [reply tests/test__collection_api.py]: ok

File: mongomock/aggregate.py (hunk 11 above)
Comment (ymoran00) [reply mongomock/aggregate.py]: ok

File: mongomock/aggregate.py (hunk 10 above)
Comment (ymoran00) [reply mongomock/aggregate.py:293]: ok

File: mongomock/aggregate.py (hunk 9 above)
Comment (ymoran00) [reply mongomock/aggregate.py:275]: ok

**Comment by ymoran00** [comment]
@pcorpet Mongo 3.7 behaves differently with negative value in $substr. Before 3.7 negative values were accepted, in 3.7 they cause errors. What should I implement then?

**Comment by pcorpet** [comment]
Let's implement the old behavior for now (to match the version of Mongo used by Travis), and add a warning when it's called that this behavior is changing.

File: tests/test__mongomock.py [hunk 13]
```diff
@@ -2275,7 +2275,25 @@ def test__aggregate33(self):
             'maxNone': {'$max': [None, None]},
             'minNone': {'$min': [None, None]},
             'avgNone': {'$avg': ['a', None]},
-            'sumNone': {'$sum': ['a', None]},
+            'sumNone': {'$sum': ['a', None]}
```
Comment (pcorpet) [inline tests/test__mongomock.py]: Nit - let's keep the trailing comma here please.

**Comment by pcorpet** [comment]
Thanks for all the changes. Please make sure the CI passes before I merge this.

**Comment by ymoran00** [comment]
> Let's implement the old behavior for now (to match the version of Mongo used by Travis), and add a warning when it's called that this behavior is changing.

Unit testing the old behavior in test__mongomock causes an error when the tox tests mongo 3.7... Maybe I'll remove unit testing for negative values for substr?

**Comment by pcorpet** [comment]
Yes, the test__mongomock is not really a unit test (it actually starts a MongoDB server in the back and compares the behaviors). Keep the unit tests in test__collection_api to be similar to the new behavior and do not test it in test__mongomock at all.

File: mongomock/aggregate.py [hunk 14]
```diff
@@ -5,6 +5,8 @@
 import copy
 import datetime
 import itertools
+import warnings
```
Comment (pcorpet) [inline mongomock/aggregate.py]: Please move after the other std library imports (after random but before six), and move the six imports (third party) as a different import group.

**Review by pcorpet — APPROVED** [review]
